packages/back-end/src/services/safeRolloutTimeSeries.ts

```diff
@@ -0,0 +1,210 @@
+import md5 from "md5";
+import { isFactMetricId, expandMetricGroups } from "shared/experiments";
+import { ReqContext } from "back-end/types/organization";
+import {
+  FactMetricInterface,
+  FactTableInterface,
+} from "back-end/types/fact-table";
+import { getFactTableMap } from "back-end/src/models/FactTableModel";
+import { SafeRolloutInterface } from "back-end/src/validators/safe-rollout";
+import {
+  SafeRolloutSnapshotInterface,
+  SafeRolloutSnapshotSettings,
+  MetricForSafeRolloutSnapshot,
+  SafeRolloutSnapshotAnalysisSettings,
+  SafeRolloutSnapshotMetricInterface,
+} from "back-end/src/validators/safe-rollout-snapshot";
+import {
+  CreateMetricTimeSeriesSingleDataPoint,
+  MetricTimeSeriesValue,
+  MetricTimeSeriesVariation,
+} from "back-end/src/validators/metric-time-series";
+import { logger } from "back-end/src/util/logger";
+
+export async function updateSafeRolloutTimeSeries({
+  context,
+  safeRollout,
+  safeRolloutSnapshot,
+}: {
+  context: ReqContext;
+  safeRollout: SafeRolloutInterface;
+  safeRolloutSnapshot: SafeRolloutSnapshotInterface;
+}) {
+  // TODO: Confirm if we want the same behavior here initially
+  if (
+    // Dimensioned safe rollouts are not supported
+    (safeRolloutSnapshot.dimension !== "" &&
+      safeRolloutSnapshot.dimension !== undefined) ||
+    // And no way of generating a data point if there are no metrics monitored, but shouldn't happen
+    safeRollout.guardrailMetricIds.length === 0
+  ) {
+    return;
+  }
+
+  const metricGroups = await context.models.metricGroups.getAll();
+  const metricsIds = expandMetricGroups(
+    safeRollout.guardrailMetricIds,
+    metricGroups
+  );
+
+  // TODO: Do we have more than one analysis type for safeRollout?
+  const analysis = safeRolloutSnapshot.analyses?.[0];
+  const analysisResults = analysis?.results?.[0];
+  const variations = analysisResults?.variations;
+
+  if (!variations || variations.length === 0) {
+    return;
+  }
+
+  // Only control & variant are expected for Safe Rollouts
+  if (variations.length !== 2) {
+    logger.warn(
+      `Safe Rollout ${safeRollout.id} has ${variations.length} variations, expected 2`
+    );
+  }
+
+  let factMetrics: FactMetricInterface[] | undefined = undefined;
+  let factTableMap: Map<string, FactTableInterface> | undefined = undefined;
+  const factMetricsIds: string[] = metricsIds.filter(isFactMetricId);
+  if (factMetricsIds.length > 0) {
+    factMetrics = await context.models.factMetrics.getByIds(factMetricsIds);
+    factTableMap = await getFactTableMap(context);
+  }
+
+  const timeSeriesVariationsPerMetricId = metricsIds.reduce((acc, metricId) => {
+    acc[metricId] = variations.map((_, variationIndex) => ({
+      id: safeRolloutSnapshot.settings.variations[variationIndex].id,
+      name: variationIndex === 0 ? "control" : "variation",
```

Comment:
Do we have these names hard coded somewhere in the safe rollout interface? I couldn't find in a brief check.